core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -326,21 +351,22 @@ V put(K key, V value, Function<ByteBuffer, V> transformer) {
             ei = c.allocateEntryAndKey(key);
             if (ei == -1) {
                 rebalance(c);
-                put(key, value, transformer);
-                return null;
+                return put(key, value, transformer);
             }
-            int prevEi = c.linkEntry(ei, true, key);
+            int prevEi = c.linkEntry(ei, key);
             if (prevEi != ei) {
                 ei = prevEi;
                 prevHi = c.getHandleIndex(prevEi);
+                if (prevHi != -1) {
```

Comment:
Please add a comment here, or at the beginning of the put, that we can swap the pointer to handle from one value to another only when we sure that previous handle was deleted (or its index was -1). This is an important invariant...